// @moonwell-fi/moonwell-sdk/_cjs/actions/governance/getStakingSnapshots.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStakingSnapshots = void 0;
const axios_1 = require("axios");
const index_js_1 = require("../../common/index.js");
async function getStakingSnapshots(client, args) {
    const environment = (0, index_js_1.getEnvironmentFromArgs)(client, args);
    if (!environment) {
        return [];
    }
    try {
        const response = await axios_1.default.post(environment.indexerUrl, {
            query: `
          query {
            stakingDailySnapshots(
              limit: 365,
              orderBy: "timestamp"
              orderDirection: "desc"
              where: {chainId: ${environment.chainId}}
            ) {
              items {
                chainId
                totalStaked
                totalStakedUSD
                timestamp
              }
            }
          }
        `,
        });
        if (response.status === 200 && response.data?.data?.stakingDailySnapshots) {
            return response.data?.data?.stakingDailySnapshots.items;
        }
        else {
            return [];
        }
    }
    catch (ex) {
        console.error("An error occured while fetching getStakingSnapshots...", ex);
        return [];
    }
}
exports.getStakingSnapshots = getStakingSnapshots;
