// @moonwell-fi/moonwell-sdk/actions/core/user-rewards/getUserRewards.ts
import type { Address } from "viem";
import type { MoonwellClient } from "../../../client/createMoonwellClient.js";
import { getEnvironmentsFromArgs } from "../../../common/index.js";
import type { NetworkParameterType } from "../../../common/types.js";
import type { Chain } from "../../../environments/index.js";
import type { UserMarketReward } from "../../../types/userReward.js";
import { getUserRewardsData } from "./common.js";

export type GetUserRewardsParameters<
  environments,
  network extends Chain | undefined,
> = NetworkParameterType<environments, network> & {
  /** User address*/
  userAddress: Address;
};

export type GetUserRewardsReturnType = Promise<UserMarketReward[]>;

export async function getUserRewards<
  environments,
  Network extends Chain | undefined,
>(
  client: MoonwellClient,
  args: GetUserRewardsParameters<environments, Network>,
): GetUserRewardsReturnType {
  const { userAddress } = args;

  const environments = getEnvironmentsFromArgs(client, args);

  const result = await Promise.all(
    environments.map((environment) =>
      getUserRewardsData({
        environment,
        account: userAddress,
      }),
    ),
  );

  return result.flat();
}
