// @moonwell-fi/moonwell-sdk/_cjs/client/createActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createActions = void 0;
const getMarket_js_1 = require("../actions/core/markets/getMarket.js");
const getMarkets_js_1 = require("../actions/core/markets/getMarkets.js");
const getUserBalances_js_1 = require("../actions/core/getUserBalances.js");
const getUserPosition_js_1 = require("../actions/core/user-positions/getUserPosition.js");
const getUserPositions_js_1 = require("../actions/core/user-positions/getUserPositions.js");
const getUserReward_js_1 = require("../actions/core/user-rewards/getUserReward.js");
const getUserRewards_js_1 = require("../actions/core/user-rewards/getUserRewards.js");
const getDelegates_js_1 = require("../actions/governance/getDelegates.js");
const getDiscussions_js_1 = require("../actions/governance/getDiscussions.js");
const getGovernanceTokenInfo_js_1 = require("../actions/governance/getGovernanceTokenInfo.js");
const getStakingInfo_js_1 = require("../actions/governance/getStakingInfo.js");
const getStakingSnapshots_js_1 = require("../actions/governance/getStakingSnapshots.js");
const getUserStakingInfo_js_1 = require("../actions/governance/getUserStakingInfo.js");
const getUserVoteReceipt_js_1 = require("../actions/governance/getUserVoteReceipt.js");
const getUserVotingPowers_js_1 = require("../actions/governance/getUserVotingPowers.js");
const getProposal_js_1 = require("../actions/governance/proposals/getProposal.js");
const getProposals_js_1 = require("../actions/governance/proposals/getProposals.js");
const getSnapshotProposal_js_1 = require("../actions/governance/snapshot/getSnapshotProposal.js");
const getSnapshotProposals_js_1 = require("../actions/governance/snapshot/getSnapshotProposals.js");
const getMorphoUserBalances_js_1 = require("../actions/morpho/getMorphoUserBalances.js");
const getMorphoMarket_js_1 = require("../actions/morpho/markets/getMorphoMarket.js");
const getMorphoMarkets_js_1 = require("../actions/morpho/markets/getMorphoMarkets.js");
const getMorphoMarketUserPosition_js_1 = require("../actions/morpho/user-positions/getMorphoMarketUserPosition.js");
const getMorphoMarketUserPositions_js_1 = require("../actions/morpho/user-positions/getMorphoMarketUserPositions.js");
const getMorphoVaultUserPosition_js_1 = require("../actions/morpho/user-positions/getMorphoVaultUserPosition.js");
const getMorphoVaultUserPositions_js_1 = require("../actions/morpho/user-positions/getMorphoVaultUserPositions.js");
const getMorphoUserRewards_js_1 = require("../actions/morpho/user-rewards/getMorphoUserRewards.js");
const getMorphoVault_js_1 = require("../actions/morpho/vaults/getMorphoVault.js");
const getMorphoVaults_js_1 = require("../actions/morpho/vaults/getMorphoVaults.js");
const createActions = (client) => {
    return {
        getMarket: (args) => (0, getMarket_js_1.getMarket)(client, args),
        getMarkets: (args) => (0, getMarkets_js_1.getMarkets)(client, args),
        getUserPosition: (args) => (0, getUserPosition_js_1.getUserPosition)(client, args),
        getUserPositions: (args) => (0, getUserPositions_js_1.getUserPositions)(client, args),
        getUserReward: (args) => (0, getUserReward_js_1.getUserReward)(client, args),
        getUserRewards: (args) => (0, getUserRewards_js_1.getUserRewards)(client, args),
        getUserBalances: (args) => (0, getUserBalances_js_1.getUserBalances)(client, args),
        getProposal: (args) => (0, getProposal_js_1.getProposal)(client, args),
        getProposals: (args) => (0, getProposals_js_1.getProposals)(client, args),
        getSnapshotProposal: (args) => (0, getSnapshotProposal_js_1.getSnapshotProposal)(client, args),
        getSnapshotProposals: (args) => (0, getSnapshotProposals_js_1.getSnapshotProposals)(client, args),
        getDelegates: () => (0, getDelegates_js_1.getDelegates)(client),
        getDiscussions: () => (0, getDiscussions_js_1.getDiscussions)(client),
        getGovernanceTokenInfo: (args) => (0, getGovernanceTokenInfo_js_1.getGovernanceTokenInfo)(client, args),
        getStakingInfo: (args) => (0, getStakingInfo_js_1.getStakingInfo)(client, args),
        getStakingSnapshots: (args) => (0, getStakingSnapshots_js_1.getStakingSnapshots)(client, args),
        getUserStakingInfo: (args) => (0, getUserStakingInfo_js_1.getUserStakingInfo)(client, args),
        getUserVoteReceipt: (args) => (0, getUserVoteReceipt_js_1.getUserVoteReceipt)(client, args),
        getUserVotingPowers: (args) => (0, getUserVotingPowers_js_1.getUserVotingPowers)(client, args),
        getMorphoMarket: (args) => (0, getMorphoMarket_js_1.getMorphoMarket)(client, args),
        getMorphoMarkets: (args) => (0, getMorphoMarkets_js_1.getMorphoMarkets)(client, args),
        getMorphoMarketUserPosition: (args) => (0, getMorphoMarketUserPosition_js_1.getMorphoMarketUserPosition)(client, args),
        getMorphoMarketUserPositions: (args) => (0, getMorphoMarketUserPositions_js_1.getMorphoMarketUserPositions)(client, args),
        getMorphoVaultUserPosition: (args) => (0, getMorphoVaultUserPosition_js_1.getMorphoVaultUserPosition)(client, args),
        getMorphoVaultUserPositions: (args) => (0, getMorphoVaultUserPositions_js_1.getMorphoVaultUserPositions)(client, args),
        getMorphoUserRewards: (args) => (0, getMorphoUserRewards_js_1.getMorphoUserRewards)(client, args),
        getMorphoVault: (args) => (0, getMorphoVault_js_1.getMorphoVault)(client, args),
        getMorphoVaults: (args) => (0, getMorphoVaults_js_1.getMorphoVaults)(client, args),
        getMorphoUserBalances: (args) => (0, getMorphoUserBalances_js_1.getMorphoUserBalances)(client, args),
    };
};
exports.createActions = createActions;
