// @moonwell-fi/moonwell-sdk/_esm/client/createActions.js
import { getMarket, } from "../actions/core/markets/getMarket.js";
import { getMarkets, } from "../actions/core/markets/getMarkets.js";
import { getUserBalances, } from "../actions/core/getUserBalances.js";
import { getUserPosition, } from "../actions/core/user-positions/getUserPosition.js";
import { getUserPositions, } from "../actions/core/user-positions/getUserPositions.js";
import { getUserReward, } from "../actions/core/user-rewards/getUserReward.js";
import { getUserRewards, } from "../actions/core/user-rewards/getUserRewards.js";
import { getDelegates, } from "../actions/governance/getDelegates.js";
import { getDiscussions, } from "../actions/governance/getDiscussions.js";
import { getGovernanceTokenInfo, } from "../actions/governance/getGovernanceTokenInfo.js";
import { getStakingInfo, } from "../actions/governance/getStakingInfo.js";
import { getStakingSnapshots, } from "../actions/governance/getStakingSnapshots.js";
import { getUserStakingInfo, } from "../actions/governance/getUserStakingInfo.js";
import { getUserVoteReceipt, } from "../actions/governance/getUserVoteReceipt.js";
import { getUserVotingPowers, } from "../actions/governance/getUserVotingPowers.js";
import { getProposal, } from "../actions/governance/proposals/getProposal.js";
import { getProposals, } from "../actions/governance/proposals/getProposals.js";
import { getSnapshotProposal, } from "../actions/governance/snapshot/getSnapshotProposal.js";
import { getSnapshotProposals, } from "../actions/governance/snapshot/getSnapshotProposals.js";
import { getMorphoUserBalances } from "../actions/morpho/getMorphoUserBalances.js";
import { getMorphoMarket } from "../actions/morpho/markets/getMorphoMarket.js";
import { getMorphoMarkets } from "../actions/morpho/markets/getMorphoMarkets.js";
import { getMorphoMarketUserPosition } from "../actions/morpho/user-positions/getMorphoMarketUserPosition.js";
import { getMorphoMarketUserPositions } from "../actions/morpho/user-positions/getMorphoMarketUserPositions.js";
import { getMorphoVaultUserPosition } from "../actions/morpho/user-positions/getMorphoVaultUserPosition.js";
import { getMorphoVaultUserPositions } from "../actions/morpho/user-positions/getMorphoVaultUserPositions.js";
import { getMorphoUserRewards } from "../actions/morpho/user-rewards/getMorphoUserRewards.js";
import { getMorphoVault } from "../actions/morpho/vaults/getMorphoVault.js";
import { getMorphoVaults } from "../actions/morpho/vaults/getMorphoVaults.js";
export const createActions = (client) => {
    return {
        getMarket: (args) => getMarket(client, args),
        getMarkets: (args) => getMarkets(client, args),
        getUserPosition: (args) => getUserPosition(client, args),
        getUserPositions: (args) => getUserPositions(client, args),
        getUserReward: (args) => getUserReward(client, args),
        getUserRewards: (args) => getUserRewards(client, args),
        getUserBalances: (args) => getUserBalances(client, args),
        getProposal: (args) => getProposal(client, args),
        getProposals: (args) => getProposals(client, args),
        getSnapshotProposal: (args) => getSnapshotProposal(client, args),
        getSnapshotProposals: (args) => getSnapshotProposals(client, args),
        getDelegates: () => getDelegates(client),
        getDiscussions: () => getDiscussions(client),
        getGovernanceTokenInfo: (args) => getGovernanceTokenInfo(client, args),
        getStakingInfo: (args) => getStakingInfo(client, args),
        getStakingSnapshots: (args) => getStakingSnapshots(client, args),
        getUserStakingInfo: (args) => getUserStakingInfo(client, args),
        getUserVoteReceipt: (args) => getUserVoteReceipt(client, args),
        getUserVotingPowers: (args) => getUserVotingPowers(client, args),
        getMorphoMarket: (args) => getMorphoMarket(client, args),
        getMorphoMarkets: (args) => getMorphoMarkets(client, args),
        getMorphoMarketUserPosition: (args) => getMorphoMarketUserPosition(client, args),
        getMorphoMarketUserPositions: (args) => getMorphoMarketUserPositions(client, args),
        getMorphoVaultUserPosition: (args) => getMorphoVaultUserPosition(client, args),
        getMorphoVaultUserPositions: (args) => getMorphoVaultUserPositions(client, args),
        getMorphoUserRewards: (args) => getMorphoUserRewards(client, args),
        getMorphoVault: (args) => getMorphoVault(client, args),
        getMorphoVaults: (args) => getMorphoVaults(client, args),
        getMorphoUserBalances: (args) => getMorphoUserBalances(client, args),
    };
};
